### PR TITLE
[SPARK-8559][MLlib] Support Association Rule Generation

### PR DESCRIPTION
**Pull request by feynmanliang**
Distributed generation of single-consequent association rules from a RDD of frequent itemsets. Tests referenced against `R`'s implementation of A Priori in [arules](http://cran.r-project.org/web/packages/arules/index.html).

### CODE REVIEW

**Comment by sujkh85** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-8559][MLlib] Support Association Rule Generation (#7005)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by mengxr** [comment]
@feynmanliang Please check Java compatibility by providing a Java unit test.

**Comment by feynmanliang** [comment]
@mengxr Done, although it feels hacky.

 * Is it okay to Scala `ClassTag` for initializing `FPGrowthModel` in Java (see [here](https://github.com/feynmanliang/spark/commit/0cc1a6a61cde78bb24eb2890b99f87c4a491f948#diff-2a23751f47a8b42138166dd3614af248R59)) solely for testing purposes since real use cases will get `FPGrowthModel` directly from `FPGrowth`?
 * I created a separate [`runJava`](https://github.com/feynmanliang/spark/commit/0cc1a6a61cde78bb24eb2890b99f87c4a491f948#diff-d9f14fbcb9174c0c811a8cf1fe099778R89) method for running `AssociationRules` from Java since I couldn't figure out how to get recursion to call the other method when provided an implicit `fakeClassTag`. Is there a better solution?

**Comment by feynmanliang** [comment]
Another thought: `FPGrowthModel` is a result type for frequent itemsets and has nothing specific to the `FPGrowth` algorithm itself. Should we consider extracting/renaming/generalizing the class for other frequent itemset algorithms (e.g. distributed A Priori) to code againt?

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.mllib.fpm
+
+import java.lang.{Iterable => JavaIterable}
+
+import org.apache.spark.api.java.JavaSparkContext.fakeClassTag
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: organize imports

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ * association rules which have a single item as the consequent.
+ */
+@Experimental
+class AssociationRules private (
+    private var minConfidence: Double) extends Logging with Serializable {
+
+  /**
+   * Constructs a default instance with default parameters {minConfidence = 0.5}.
+   */
+  def this() = this(0.5)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: Choose a higher value as default. R uses 0.8.

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        items.partition(_ == item) match {
+          case (consequent, antecedent) => (antecedent.toList, (consequent.toList, itemset.freq))
+        }
+      } :+ (items.toList, (Nil, itemset.freq))
+    }.aggregateByKey(Map[List[Item], Long]().empty)(
+      { case (acc, (consequent, freq)) => acc + (consequent -> freq) },
+      { (acc1, acc2) => acc1 ++ acc2 }
+    ).flatMap { case (antecedent, consequentToFreq) =>
+      consequentToFreq.flatMap { case (consequent, freqUnion) =>
+        val freqAntecedent = if (antecedent == Nil) numTransactions else consequentToFreq(Nil)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: Is `() => { ... }` really useful? We should check literature to confirm it.

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  def run[Item: ClassTag](model: FPGrowthModel[Item]): RDD[Rule[Item]] = {
+    val freqItemsets = model.freqItemsets
+    val numTransactions = model.numTransactions
+
+    freqItemsets.flatMap { itemset =>
+      val items = itemset.items
+      items.map { item =>
+        // Key using List[Item] because cannot use Array[Item] for map-side combining
+        items.partition(_ == item) match {
+          case (consequent, antecedent) => (antecedent.toList, (consequent.toList, itemset.freq))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: `toList` -> `toSeq`

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        }
+      } :+ (items.toList, (Nil, itemset.freq))
+    }.aggregateByKey(Map[List[Item], Long]().empty)(
+      { case (acc, (consequent, freq)) => acc + (consequent -> freq) },
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]:
~~~
seqOp = { case (acc, ...
} ...
~~~

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      { (acc1, acc2) => acc1 ++ acc2 }
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: `++`

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    ).flatMap { case (antecedent, consequentToFreq) =>
+      consequentToFreq.flatMap { case (consequent, freqUnion) =>
+        val freqAntecedent = if (antecedent == Nil) numTransactions else consequentToFreq(Nil)
+        val confidence = freqUnion.toDouble / freqAntecedent.toDouble
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: the second `toDouble` is not necessary

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+   * @param antecedent hypotheses of the rule
+   * @param consequent conclusion of the rule
+   * @param confidence the confidence of the rule
+   * @tparam Item item type
+   */
+  @Experimental
+  case class Rule[Item: ClassTag](
+      antecedent: Array[Item],
+      consequent: Array[Item],
+      confidence: Double)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: add frequency for the union and antecedent. move confidence to a method.

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+   * :: Experimental ::
+   *
+   * An association rule between sets of items.
+   * @param antecedent hypotheses of the rule
+   * @param consequent conclusion of the rule
+   * @param confidence the confidence of the rule
+   * @tparam Item item type
+   */
+  @Experimental
+  case class Rule[Item: ClassTag](
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: remove `case`, `ClassTag` and make constructor private

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        if (!consequent.isEmpty && confidence >= minConfidence) {
+          Some(Rule[Item](antecedent.toArray, consequent.toArray, confidence))
+        } else {
+          None
+        }
+      }
+    }
+  }
+
+  def runJava[Item](model: FPGrowthModel[Item]): JavaRDD[Rule[Item]] = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: If we can take `JavaRDD[FreqItemset]`, we can keep the name as `run`, similar to `FPGrowth`

File: mllib/src/test/scala/org/apache/spark/mllib/fpm/AssociationRulesSuite.scala [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    val numTransactions = 6
+    val freqItemsets = sc.parallelize(Seq(
+      (Set("s"), 3L), (Set("z"), 5L), (Set("x"), 4L), (Set("t"), 3L), (Set("y"), 3L),
+      (Set("r"), 3L),
+      (Set("x", "z"), 3L), (Set("t", "y"), 3L), (Set("t", "x"), 3L), (Set("s", "x"), 3L),
+      (Set("y", "x"), 3L), (Set("y", "z"), 3L), (Set("t", "z"), 3L),
+      (Set("y", "x", "z"), 3L), (Set("t", "x", "z"), 3L), (Set("t", "y", "z"), 3L),
+      (Set("t", "y", "x"), 3L),
+      (Set("t", "y", "x", "z"), 3L))
+      .map({ case (items, freq) => new FPGrowth.FreqItemset(items.toArray, freq) }))
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/fpm/AssociationRulesSuite.scala]:
~~~
.map { case (items, freq) =>
  new FPGrowth.FreqItemset(items.toArray, freq)
})
~~~

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+      val items = itemset.items
+      items.flatMap { item =>
+        items.partition(_ == item) match {
+          // Itemsets and items in itemsets are unique, so every (antecedent, consequent) is unique
+          case (consequent, antecedent) if !antecedent.isEmpty =>
+            Some((antecedent.toSeq, (consequent.toSeq, itemset.freq)))
+          case _ => None
+        }
+      } :+ (items.toSeq, (Nil, itemset.freq))
+    }.aggregateByKey(Map[Seq[Item], Long]().empty)(
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: Just realized that this step is not necessary. After `flatMap`, we only need to join back with `freqItemsets`:

~~~scala
val candidates = freqItemsets.flatMap { ... }
candidates.join(freqItemsets.map(x => (x.items.toSeq, x.freq))
  .map { case (antecendent, ((consequent, freqUnion), freqAntecedent))) =>
    new Rule(antecendent.toArray, consequent.toArray, freqUnion, freqAntecedent)
  }.filter(_.confidence >= minConfidence)
}
~~~

It will be the same amount of shuffle.

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+   * @param antecedent hypotheses of the rule
+   * @param consequent conclusion of the rule
+   * @param freqUnion the frequency (num. occurrences) of the union of the antecedent and consequent
+   *                  as subsets of transactions in the transaction data
+   * @param freqAntecedent the frequency of the antecedent in the transaction data
+   * @tparam Item item type
+   */
+  @Experimental
+  class Rule[Item] private[mllib] (
+      antecedent: Array[Item],
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: Add `val` to make them publicly accessible. This could be checked by generating the API doc.

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+      consequent: Array[Item],
+      freqUnion: Double,
+      freqAntecedent: Double) extends Serializable {
+
+    def confidence: Double = 1.0 * freqUnion / freqAntecedent
+
+    require(antecedent.toSet.intersect(consequent.toSet).isEmpty, {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: minor: This could be quite expensive to check. It generates many temporary sets when we have many rules. Given the fact that `Rule` can be only created by `AssociationRules`, we could move this to unit test if it hurts the performance. But let's keep it here for now.

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/FPGrowth.scala [hunk 15]
```diff
@@ -40,7 +40,14 @@ import org.apache.spark.storage.StorageLevel
  * @tparam Item item type
  */
 @Experimental
-class FPGrowthModel[Item: ClassTag](val freqItemsets: RDD[FreqItemset[Item]]) extends Serializable
+class FPGrowthModel[Item: ClassTag](
+    val freqItemsets: RDD[FreqItemset[Item]],
+    val numTransactions: Long) extends Serializable {
+
+  def this(javaFreqItemsets: JavaRDD[FreqItemset[Item]], javaNumTransactions: Long) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/FPGrowth.scala]: Let's do this in a separate PR. We need a Java unit test to make sure it works. The `numTransactions` change could be in a separate PR as well (with unit test).

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala (hunk 13 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: OK.

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala (hunk 14 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/FPGrowth.scala (hunk 15 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/FPGrowth.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala (hunk 12 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala]: OK.

**Comment by mengxr** [comment]
LGTM. Merged into master. Please create JIRAs for Python API and user guide. Another useful method to add might be `FPGrowthModel.generateAssociationRules(confidence: Double)`.
